apps/emqx_gcp_device/src/emqx_gcp_device_authn.erl

```diff
@@ -0,0 +1,233 @@
+    ] ++ emqx_authn_schema:common_fields().
+
+refs(#{<<"mechanism">> := <<"gcp_device">>}) ->
+    [hoconsc:ref(?MODULE, gcp_device)];
+refs(_) ->
+    throw(#{
+        field_name => mechanism,
+        expected => "gcp_device"
+    }).
+
+% The check logic is the following:
+%% 1. If clientid is not GCP-like or password is not a JWT, the result is ignore
+%% 2. If clientid is GCP-like and password is a JWT, but expired, the result is password_error
+%% 3. If clientid is GCP-like and password is a valid and not expired JWT:
+%%  3.1 If there are no keys for the client, the result is ignore
+%%  3.2 If there are some keys for the client:
+%%   3.2.1 If there are no actual (not expired keys), the result is password_error
+%%   3.2.2 If there are some actual keys and one of them matches the JWT, the result is success
+%%   3.2.3 If there are some actual keys and none of them matches the JWT, the result is password_error
+check(#{password := Password} = ClientInfo) ->
+    case gcp_deviceid_from_clientid(ClientInfo) of
+        {ok, DeviceId} ->
+            case is_valid_jwt(Password) of
+                true ->
+                    check_jwt(ClientInfo, DeviceId);
+                {false, not_a_jwt} ->
+                    ?tp(authn_gcp_device_check, #{
+                        result => ignore, reason => "not a JWT", client => ClientInfo
+                    }),
+                    ?TRACE_AUTHN_PROVIDER(debug, "auth_ignored", #{
+                        reason => "not a JWT",
+                        client => ClientInfo
+                    }),
+                    ignore;
+                {false, expired} ->
+                    ?tp(authn_gcp_device_check, #{
+                        result => not_authorized, reason => "expired JWT", client => ClientInfo
+                    }),
+                    ?TRACE_AUTHN_PROVIDER(info, "auth_failed", #{
+                        reason => "expired JWT",
+                        client => ClientInfo
+                    }),
+                    {error, not_authorized}
+            end;
+        not_a_gcp_clientid ->
+            ?tp(authn_gcp_device_check, #{
+                result => ignore, reason => "not a GCP ClientId", client => ClientInfo
+            }),
+            ?TRACE_AUTHN_PROVIDER(debug, "auth_ignored", #{
+                reason => "not a GCP ClientId",
+                client => ClientInfo
+            }),
+            ignore
+    end.
+
+check_jwt(ClientInfo, DeviceId) ->
+    case emqx_gcp_device:get_device_actual_keys(DeviceId) of
+        not_found ->
+            ?tp(authn_gcp_device_check, #{
+                result => ignore, reason => "key not found", client => ClientInfo
+            }),
+            ?TRACE_AUTHN_PROVIDER(debug, "auth_ignored", #{
+                reason => "key not found",
+                client => ClientInfo
+            }),
+            ignore;
+        Keys ->
+            case any_key_matches(Keys, ClientInfo) of
```

Comment:
Q: is it possible that `[] = Keys`?  If so, wouldn't it fall under case 3.1 of the comment above (result should be `ignore`)?

Reply:
`Keys = []` is considered to be `3.2.1` case